packages/remirror__extension-annotation/src/annotation-plugin.ts

```diff
@@ -24,7 +24,92 @@ export class AnnotationState<Type extends Annotation = Annotation> {
    */
   decorationSet = DecorationSet.empty;
 
-  constructor(private readonly getStyle: GetStyle<Type>) {}
+  constructor(
+    private readonly getStyle: GetStyle<Type>,
+    private readonly map: MapLike<string, OmitText<Type>>,
+    private readonly transformPosition: (pos: number) => number,
+    private readonly transformPositionBeforeRender: (pos: number) => number | null,
+  ) {}
+
+  addAnnotation(addAction: AddAnnotationAction<Type>): void {
+    const { id } = addAction.annotationData;
+    this.map.set(id, {
+      ...addAction.annotationData,
+      from: this.transformPosition(addAction.from),
+      to: this.transformPosition(addAction.to),
+    } as OmitText<Type>);
+  }
+
+  updateAnnotation(updateAction: UpdateAnnotationAction<Type>): void {
+    assert(this.map.has(updateAction.annotationId));
+
+    this.map.set(updateAction.annotationId, {
+      ...this.map.get(updateAction.annotationId),
+      ...updateAction.annotationData,
+    } as OmitText<Type>);
+  }
+
+  removeAnnotations(removeAction: RemoveAnnotationsAction): void {
+    removeAction.annotationIds.forEach((id) => {
+      this.map.delete(id);
+    });
+  }
+
+  setAnnotations(setAction: SetAnnotationsAction<Type>): void {
+    // YJS maps don't support clear
+    this.map.clear?.();
```

Comment:
The `?.` should be redundant once https://github.com/yjs/yjs/pull/309 is released